### PR TITLE
Fixed the SHM segment smashing while reading with security <2.0.x> [10165]

### PR DESCRIPTION
**Pull request by rsanchez15**
This PR includes:
- Fix the shared memory segment writing access while reading when the RTPS message and sub-messages are encrypted.
- Communication tests

### CODE REVIEW

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3335/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/543/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2992/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3367/)

**Review by MiguelCompany — APPROVED** [review]

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3347/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/552/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3001/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3379/)

File: src/cpp/rtps/messages/MessageReceiver.cpp [hunk 1]
```diff
@@ -197,10 +198,13 @@ void MessageReceiver::processCDRMsg(
 
     if (decode_ret == 0)
     {
-        // Swap
-        std::swap(msg, auxiliary_buffer);
+        // The original CDRMessage buffer (msg) now points to the proprietary temporary buffer crypto_msg_.
+        // The auxiliary buffer now points to the propietary temporary buffer crypto_submsg_.
+        // This way each decoded submessage will be process using the crypto_submsg_ buffer.
```
Comment (MiguelCompany) [inline src/cpp/rtps/messages/MessageReceiver.cpp:203]:
```suggestion
        // This way each decoded sub-message will be processed using the crypto_submsg_ buffer.
```

File: src/cpp/rtps/messages/MessageReceiver.cpp [hunk 2]
```diff
@@ -1161,6 +1162,6 @@ bool MessageReceiver::proc_Submsg_HeartbeatFrag(
     return true;
 }
 
-}
+} // namespace rtps
 } /* namespace rtps */
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/rtps/messages/MessageReceiver.cpp:1169]: This one is missing again

**Review by MiguelCompany — CHANGES_REQUESTED** [review]
Two minor details only

**Review by MiguelCompany — APPROVED** [review]
LGTM

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3358/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/561/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3007/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3390/)

**Pull request merged**